canonical/mongodb-operator
# Add pbm configs

## Summary
backups to s3 require options from the user. As discussed s3 backups should be performed in this way:
```
juju deploy s3-integrator 
juju run-action s3-integrator/leader sync-s3-credentials \
    access-key=<your_key> \
    secret-key=<your_secret_key>
juju config s3-integrator \
    bucket=s3-bucket-name \
    path-prefix=some-dir \
    access-key=login \
    secret-key=password
juju relate s3-integrator mysql
juju run-action --wait mysql-k8s/0 create-backup
juju run-action --wait mysql-k8s/0 list-backups
juju run-action --wait mysql-k8s/0 restore backup-id=<%Y-%m-%d-%H-%M>
```
so this PR:
1.  implements the relation for  the s3 integrator 
2. checks credentials are valid
3. checks provided config options are valid
4. checks pbm is ready to serve requests 
5. performs resync and waits

## Context
Backups for Charmed MongoDB make use of [PBM](https://docs.percona.com/percona-backup-mongodb/index.html). PBM requires configurations in s3 such as: storage.s3.region storage.s3.bucket, storage.s3.prefix, storage.s3.credentials.access-key-id, storage.s3.credentials.secret-access-key, 

**Resync takes a long time**

## Manual test
Manually tested by listing config options with `percona-backup-mongodb config --list` to verify that `juju config mongodb  option="value"` was effective:

```
juju run-action s3-integrator/leader sync-s3-credentials access-key="access-key-user1" secret-key="secret-key-user1" --wait
juju relate s3-integrator mongodb

# verify blocked with: “s3 configurations are incompatible”  in juju status.

juju config s3-integrator \
    path="data/pbm/backup" \
    region="us-west-2" \
    bucket="pbm-test-bucket-1" 

# verify active state in juju status.

juju run-action mongodb/0 create-backup --wait


juju run-action s3-integrator/leader sync-s3-credentials access-key="access-key-user2" secret-key="secret-key-user2" --wait
juju config s3-integrator \
    bucket="pbm-test-bucket-3" 

# verify waiting with “waiting to sync s3 configurations” in juju status

juju run-action mongodb/0 create-backup --wait

# now check the buckets for the backups in AWS
```

## Future PRs
As mentioned this is the first PR of several to get the backup action up and running. The rest are:
- Add unit tests
- Integration tests
